amd_tressfx_vulkan/src/TressFXRendererVulkan.cpp

```diff
@@ -0,0 +1,1569 @@
+                               VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC,
+                               &bufferDescriptor),
+            getWriteDescriptor(m_pass2_set, IDSRV_HEAD_PPLL,
+                               VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, &headPLLDescriptor),
+            getWriteDescriptor(m_pass2_set, IDSRV_PPLL, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER,
+                               &PPLL_Buffer_Descriptor),
+            getWriteDescriptor(m_pass2_set, IDSRV_HAIRSM,
+                               VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, &shadowDescriptor),
+            getWriteDescriptor(m_shadow_pass_set, IDSRV_CONSTANTS_BUFFER,
+                               VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC,
+                               &bufferDescriptor),
+        };
+
+        vkUpdateDescriptorSets(pvkDevice, AMD_ARRAY_SIZE(writeDescriptorSets),
+                               writeDescriptorSets, 0, nullptr);
+    }
+    else
+    {
+        VkWriteDescriptorSet writeDescriptorSets[] = {
+            getWriteDescriptor(m_shadow_pass_set, IDSRV_CONSTANTS_BUFFER,
+                               VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC,
+                               &bufferDescriptor),
+        };
+
+        vkUpdateDescriptorSets(pvkDevice, AMD_ARRAY_SIZE(writeDescriptorSets),
+                               writeDescriptorSets, 0, nullptr);
+    }
+
+    return VK_SUCCESS;
+}
+
+//--------------------------------------------------------------------------------------
+//
+// CreatePPLL
+//
+// Creates the per pixel linked list buffers and views. To save space, the PPLL is
+// shared between multiple TressFXRenderer objects. The ref count gets incremented
+// when the PPLL is shared with a new TressFXRenderer object.
+//
+//--------------------------------------------------------------------------------------
+VkResult TressFXRenderer::CreatePPLL(VkDevice pvkDevice, int winWidth, int winHeight,
+                                     bool resize, uint32_t texture_memory_index)
+{
+    VkResult vr;
+
+    // see if the buffer needs to be resized or if refCount is 0
+    if ((winWidth != g_PPLBuffers.width) || (winHeight != g_PPLBuffers.height) ||
+        (g_PPLBuffers.refCount == 0))
+    {
+        // Release any previously allocated buffers
+        //        AMD_SAFE_RELEASE(g_PPLBuffers.pHeadPPLL_Buffer);
+        //        AMD_SAFE_RELEASE(g_PPLBuffers.pHeadPPLL_SRV);
+        //        AMD_SAFE_RELEASE(g_PPLBuffers.pHeadPPLL_UAV);
+        //        AMD_SAFE_RELEASE(g_PPLBuffers.pPPLL_Buffer);
+        //        AMD_SAFE_RELEASE(g_PPLBuffers.pPPLL_UAV);
+        //        AMD_SAFE_RELEASE(g_PPLBuffers.pPPLL_SRV);
+
+        // linked list head texture
+        VkImageCreateInfo headPPLLInfo =
+            getImageCreateInfo(VK_FORMAT_R32_UINT, winWidth, winHeight,
+                               VK_IMAGE_USAGE_SAMPLED_BIT | VK_IMAGE_USAGE_STORAGE_BIT);
+        AMD_V_RETURN(vkCreateImage(pvkDevice, &headPPLLInfo, nullptr,
```

Comment:
If hardware supports it, you should move both the image and the buffer below into the same allocation.